src/test/java/com/google/sps/managers/DatastoreManagerTest.java

```diff
@@ -0,0 +1,301 @@
+
+    ArrayList<String> stages = new ArrayList<String>();
+    stages.add("stage1id");
+    stages.add("stage2id");
+    
+    //Makes use of the Game class defined previously to build a single game for testing purposes
+    Game.Builder game = new Game.Builder("0001", "game1").setGameCreator("Adetowo").setStages(stages);
+    game.setGameDescription("For Testing").setNumTimesPlayed(14).setNumTimesFinished(12);
+    game.setNumStarVotes(6).setTotalStars(17).setNumDifficultyVotes(6).setTotalDifficulty(6);       
+    expected = game.build();
+
+    datastoreManager.createOrReplaceGame(expected);
+
+    actual = datastoreManager.retrieveGame("0001");
+
+    Assert.assertEquals(expected.getGameID(), actual.getGameID());
+  }
+
+  @Test
+  public void createOrReplaceStageAndRetrieveTest() {
+    Stage expected;
+    Stage actual;
+    
+    //Makes use of the Stage class and it's functions defined previously to build a single stage for testing purposes
+    Stage.Builder stage = new Stage.Builder("0001", 4);
+    stage.setKey("Test").setStartingHint("Testing").setStartingLocation(new Coordinates()).setHints(new ArrayList<Hint>());
+    expected = stage.build();
+
+    datastoreManager.createOrReplaceStage(expected);
+
+    actual = datastoreManager.retrieveStage("0001");
+
+    Assert.assertEquals(expected.getStageID(), actual.getStageID());
+  }
+
+  @Test
+  public void createOrReplaceSinglePlayerProgressAndRetrieveTest() {
+    SinglePlayerProgress expected;
+    SinglePlayerProgress actual;
+    
+    //Makes use of the SinglePlayerProgress class defined previously to build a single SinglePlayerProgress for testing purposes
+    SinglePlayerProgress.Builder progress = new SinglePlayerProgress.Builder("0000", "0001");
+    progress.setLocation(new Coordinates()).setHintsFound(new ArrayList<Integer>()).setStageID("0002");
+    expected = progress.build();
+
+    datastoreManager.createOrReplaceSinglePlayerProgress(expected);
+
+    actual = datastoreManager.retrieveSinglePlayerProgress("0000", "0001");
+
+    Assert.assertEquals(expected.getUserID(), actual.getUserID());
+  }
+
+  @Test
+  public void createOrReplaceHintAndRetrieveTest() {
+    Hint expected;
+    Hint actual;
+    
+    //Makes use of the Hint class defined previously to build a single Hint for testing purposes
+    Hint.Builder hint = new Hint.Builder("0001", 3);
+    hint.setLocation(new Coordinates()).setText("For testing");
+    expected = hint.build();
+
+    datastoreManager.createOrReplaceHint(expected);
+
+    actual = datastoreManager.retrieveHint("0001");
+
+    Assert.assertEquals(expected.getHintID(), actual.getHintID());
+  }
+
+  @Test
+  public void retrieveAllGamesTest() {
+    int expected = 10;
+    int actual;
+    ArrayList<Game> test = new ArrayList<>();
+  
+    for(int i = 0; i < 10; i++) test.add(getRandomGame());
+    actual =  test.size();
+
+    Assert.assertEquals(expected, actual);
+  }
+  
+  @Test
+  public void doesUsernameExistTest() {
```

Comment:
Also add a test for when the username doesn't exist.

Reply:
sure. already wroking on testing edge cases